File: generated_code/cpp/solution_2681_2.cpp
#include <vector>
#include <algorithm>
#include <numeric>

using namespace std;

class Solution {
public:
    long long putMarbles(vector<int>& weights, int k) {
        // If we need to divide into k bags, we need to make k - 1 splits.
        // Each split between index i and i+1 adds weights[i] + weights[i+1] to the total score.
        // The first element weights[0] and the last element weights[n-1] are always included
        // in both the minimum and maximum scores, so they cancel out in the difference.
        // Thus, the problem reduces to finding the sum of the k-1 largest pair sums 
        // minus the sum of the k-1 smallest pair sums.
        
        int n = weights.size();
        if (k == 1 || n == k) {
            return 0; // Only 1 way to distribute or 1 marble per bag, difference is 0.
        }
        
        vector<long long> pair_sums(n - 1);
        for (int i = 0; i < n - 1; ++i) {
            pair_sums[i] = (long long)weights[i] + weights[i + 1];
        }
        
        // To find the sum of the k-1 smallest and largest elements efficiently, 
        // we can sort the pair sums. Time complexity: O(N log N).
        sort(pair_sums.begin(), pair_sums.end());
        
        long long min_score_components = 0;
        long long max_score_components = 0;
        
        // Sum the smallest k-1 pairs and the largest k-1 pairs.
        for (int i = 0; i < k - 1; ++i) {
            min_score_components += pair_sums[i];
            max_score_components += pair_sums[n - 2 - i];
        }
        
        return max_score_components - min_score_components;
    }
};

File: generated_code/python/solution_2681_2.py
from typing import List

class Solution:
    def putMarbles(self, weights: List[int], k: int) -> int:
        # Time Complexity: O(N log N) where N is the length of weights (due to sorting)
        # Space Complexity: O(N) to store the pair sums
        #
        # ALGORITHM EXPLANATION:
        # Dividing the array into k bags requires making k - 1 splits.
        # If we split between index i and i + 1, the right boundary of the current bag 
        # is weights[i] and the left boundary of the next bag is weights[i + 1].
        # Thus, each split contributes weights[i] + weights[i + 1] to the total score.
        # The first element weights[0] and the last element weights[-1] are always 
        # part of the score regardless of how we partition the array.
        # 
        # To find the maximum score, we pick the k - 1 largest adjacent pair sums.
        # To find the minimum score, we pick the k - 1 smallest adjacent pair sums.
        # The difference between the max and min scores is simply the difference 
        # between the sum of the top k - 1 pairs and the sum of the bottom k - 1 pairs.
        
        n = len(weights)
        if k == 1 or n == k:
            return 0
            
        # Calculate all possible adjacent pair sums
        pair_sums = [weights[i] + weights[i + 1] for i in range(n - 1)]
        
        # Sort the pair sums to easily pick the smallest and largest ones
        pair_sums.sort()
        
        # Elements to pick
        m = k - 1
        
        # Sum of the m largest pairs minus the sum of the m smallest pairs
        # This bypasses needing to track weights[0] and weights[-1] since they cancel out
        max_score_contribution = sum(pair_sums[-m:]) if m > 0 else 0
        min_score_contribution = sum(pair_sums[:m]) if m > 0 else 0
        
        return max_score_contribution - min_score_contribution